sql: add cron configuration to TTL

otan opened the pull request:
Last 3 commits in this sequence for review.

Rest is #75840

otan committed:
sql: create scheduled job on TTL table creation

This commit contains the large majority of boilerplate needed to create
a SCHEDULE which periodically invokes a JOB. In this case we are
creating an hourly (for now) schedule to trigger a TTL job.

Release note: None

otan committed:
sql: implement basic row-level TTL job

This commit introduces the actual TTL job that will remove rows. It is a
very basic implementation, with configuration options and other
optimizations slated for future commits.

Release note: None

otan committed:
sql: allow configuration of cron duration for TTL jobs

This commit implements the `ttl_job_cron` syntax portion of TTL,
plumbing it into the scheduled job.

Currently only works for `CREATE TABLE`.

Release note: None

otan committed:
sql: persist schedule ids onto the descriptor

This is helpful for accounting as we don't have to hardcode and guess
any row-level TTL schedules.

Release note: None

otan committed:
sql: make sql.JobScheduleEnv method public

Reduces a little bit of duplicate code for re-use later.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76195)
<!-- Reviewable:end -->

otan committed:
paramparse: handle SET and RESET of ttl_job_cron

This commit introduces the proper handling of SET and RESET for
ttl_job_cron, which involves having to edit the cron expression of the
scheduled job. This requires injecting the "planner" into paramobserver
but only for the ALTER TABLE case.

Release note: None

otan committed:
ttljob: implement delete_batch_size, select_batch_size

Users can now configure the batch size of selecting and deleting for
individual tables with `ttl_select_batch_size` and
`ttl_delete_batch_size` respectively on the table storage parameter
arguments. The defaults are configurable with the
`sql.defaults.ttl_select_batch_size` and
`sql.defaults.ttl_delete_batch_size` jobs.

Release note: None

otan commented:
folding into schema change PR
